### PR TITLE
PIFO, FIFO: command `peek`

### PR DESCRIPTION
**Pull request by anshumanmohan**
At one point @paili0628 requested that the FIFO component support a `peek` command. I put this in, but did not expect-test it. 

This PR expect-tests `peek` in the FIFO component, and also adds the same functionality to the PIFO component. Finally, it expect-tests it for the PIFO component.

This PR builds atop of #1625 and should only be merged after 1625 is.

### FIFO

The commands are:
```
  [
    100,
    1,   // a peek!
    101,
    102,
    0,
    103,
    104,
    105,
    106,
    107,
    108,
    0,
    0,
    0,
    0
  ]
```

and I expect the answer:
```
  [
    100, // from peeking
    100, // from popping
    101, // ...
    102,
    103,
    104,
    0, // just the memory we passed; never got clobbered.
    0,
    0,
    0
  ]
```

### PIFO

The commands are:
```
  [
    101,
    102,
    0,
    0,
    103,
    104,
    201,
    202,
    1, // a peek!
    0,
    0,
    0,
    0,
    105,
    0
  ]
```

and I expect the answer:

```
  [
    101,
    102,
    201, // from peeking
    201, // from popping
    103,
    202,
    104,
    105,
    0,
    0
  ]
```

### CODE REVIEW

**Comment by rachitnigam** [comment]
Heads up: for ease of reviewing, I've updated the "base" branch from `master` to `pifo` (done by clicking "edit" next to the PR title). Once that branch is merged, the base will revert to `master` again.

File: calyx-py/calyx/queue_call.py [hunk 1]
```diff
@@ -96,35 +93,21 @@ def insert_main(prog, queue):
...
+                ),
+                cb.if_(  # If it was a pop or a peek, write ans to the answer list
+                    cmd_le_1[0].out,
+                    cmd_le_1[1],
+                    [  # AM: I'd like to have an additional check hereL
+                        # if err flag comes back raised,
+                        # we do not perform this write_ans or this incr_j
+                        write_ans,
+                        incr_j,
+                    ],
```
Comment (rachitnigam) [inline calyx-py/calyx/queue_call.py:110]: What's the intuition for why this cannot be in the `par` anymore?

**Review by rachitnigam — APPROVED** [review]
LGTM! One quick question left in there.

File: calyx-py/calyx/queue_call.py (hunk 1 above)
Comment (anshumanmohan) [reply calyx-py/calyx/queue_call.py:110]: Ah, so the diff makes it looks like we went from `par (foo)` to `seq (foo)`, but actually we went from:
```
par (
  ( 
    if_case_1: action; additional_action_for_cases_1_and_2 
  ), 
  ( 
    if_case_2: action; additional_action_for_cases_1_and_2 
  ), 
  ( 
    if_case_3: action; 
  )
) # par ends 
```
to:
```
action;
if_case_1_or_2: additional_action_for_cases_1_and_2
```
And I think that's simpler.

Basically the `action` (in our specific case, in `invoke`) was the same in all the parallel branches, so I lifted it. Then the `additional_action_for_cases_1_and_2`, which in our specific case is `write_ans; incr_j`, is run after that. These cannot be run in parallel: the latter step depends on `ans`, which is passed by reference (and potentially modified) during the former step.

Comment (rachitnigam) [reply calyx-py/calyx/queue_call.py:116]: Oh, gotcha! We should implement this kind of hoisting in a control simplification pass for Calyx. Maybe worth opening a short issue!

**Pull request merged**